File: typescript-webauthn-node/src/ceremony-store.ts
import { randomUUID } from "node:crypto";

import { AuthWebAuthnError } from "./error.js";

export type WebAuthnCeremonyKind = "registration" | "authentication";

export interface WebAuthnCeremony {
  readonly kind: WebAuthnCeremonyKind;
  readonly challenge: string;
  readonly expiresAtMs: number;
  readonly subjectId?: string | undefined;
}

export interface PutWebAuthnCeremonyInput {
  readonly kind: WebAuthnCeremonyKind;
  readonly challenge: string;
  readonly subjectId?: string | undefined;
}

export interface WebAuthnCeremonyStore {
  put(key: string, ceremony: PutWebAuthnCeremonyInput): void;
  consume(key: string, expectedKind: WebAuthnCeremonyKind): WebAuthnCeremony;
}

export interface InMemoryWebAuthnCeremonyStoreOptions {
  readonly ttlMs: number;
  readonly now?: (() => number) | undefined;
}

/**
 * Single-process ceremony store for tests and simple deployments.
 *
 * Distributed applications should implement `WebAuthnCeremonyStore` with an atomic, expiring
 * shared store. A ceremony is deleted before it is returned so every verification attempt is
 * single-use, including failed attempts.
 */
export class InMemoryWebAuthnCeremonyStore implements WebAuthnCeremonyStore {
  readonly #entries = new Map<string, WebAuthnCeremony>();
  readonly #ttlMs: number;
  readonly #now: () => number;

  constructor(options: InMemoryWebAuthnCeremonyStoreOptions) {
    if (!Number.isSafeInteger(options.ttlMs) || options.ttlMs <= 0) {
      throw new RangeError("ttlMs must be a positive safe integer");
    }
    this.#ttlMs = options.ttlMs;
    this.#now = options.now ?? Date.now;
  }

  put(key: string, ceremony: PutWebAuthnCeremonyInput): void {
    if (key === "") throw new RangeError("ceremony key must not be empty");
    if (ceremony.challenge === "")
      throw new RangeError("ceremony challenge must not be empty");
    const nowMs = this.#now();
    this.prune(nowMs);
    this.#entries.set(
      key,
      Object.freeze({
        ...ceremony,
        expiresAtMs: nowMs + this.#ttlMs,
      }),
    );
  }

  consume(key: string, expectedKind: WebAuthnCeremonyKind): WebAuthnCeremony {
    const ceremony = this.#entries.get(key);
    this.#entries.delete(key);
    if (ceremony === undefined) {
      throw new AuthWebAuthnError(
        "ceremony-missing",
        "WebAuthn ceremony was not found",
      );
    }
    if (ceremony.expiresAtMs <= this.#now()) {
      throw new AuthWebAuthnError(
        "ceremony-expired",
        "WebAuthn ceremony has expired",
      );
    }
    if (ceremony.kind !== expectedKind) {
      throw new AuthWebAuthnError(
        "ceremony-mismatch",
        "WebAuthn ceremony kind does not match",
      );
    }
    return ceremony;
  }

  prune(nowMs: number = this.#now()): number {
    let removed = 0;
    for (const [key, ceremony] of this.#entries) {
      if (ceremony.expiresAtMs <= nowMs) {
        this.#entries.delete(key);
        removed += 1;
      }
    }
    return removed;
  }

  clear(): void {
    this.#entries.clear();
  }

  get size(): number {
    return this.#entries.size;
  }
}

export function createWebAuthnCeremonyKey(kind: WebAuthnCeremonyKind): string {
  return `${kind}:${randomUUID()}`;
}

File: rust-webauthn/src/ceremony_store.rs
use std::collections::HashMap;
use std::error::Error;
use std::fmt::{Display, Formatter};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum WebauthnCeremonyKind {
    Registration,
    Authentication,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct WebauthnCeremony<State> {
    pub kind: WebauthnCeremonyKind,
    pub state: State,
    pub expires_at_ms: u64,
    pub subject_id: Option<String>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum WebauthnCeremonyStoreError {
    InvalidTtl,
    EmptyKey,
    ExpiryOverflow,
    Missing,
    Expired,
    KindMismatch,
}

impl Display for WebauthnCeremonyStoreError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(match self {
            Self::InvalidTtl => "ceremony TTL must be positive",
            Self::EmptyKey => "ceremony key must not be empty",
            Self::ExpiryOverflow => "ceremony expiry overflowed u64",
            Self::Missing => "WebAuthn ceremony was not found",
            Self::Expired => "WebAuthn ceremony has expired",
            Self::KindMismatch => "WebAuthn ceremony kind does not match",
        })
    }
}

impl Error for WebauthnCeremonyStoreError {}

/// Single-process, single-use ceremony-state storage for tests and simple deployments.
///
/// Distributed applications must use an atomic shared store with equivalent consume-before-verify
/// semantics. Every consume attempt removes the entry before expiry and kind checks, so even a
/// failed verification cannot replay the state.
#[derive(Clone, Debug)]
pub struct InMemoryWebauthnCeremonyStore<State> {
    ttl_ms: u64,
    entries: HashMap<String, WebauthnCeremony<State>>,
}

impl<State> InMemoryWebauthnCeremonyStore<State> {
    /// Creates an empty ceremony store with a fixed positive TTL.
    ///
    /// # Errors
    ///
    /// Returns [`WebauthnCeremonyStoreError::InvalidTtl`] when `ttl_ms` is zero.
    pub fn new(ttl_ms: u64) -> Result<Self, WebauthnCeremonyStoreError> {
        if ttl_ms == 0 {
            return Err(WebauthnCeremonyStoreError::InvalidTtl);
        }
        Ok(Self {
            ttl_ms,
            entries: HashMap::new(),
        })
    }

    /// Stores opaque server-side ceremony state until its checked expiry.
    ///
    /// # Errors
    ///
    /// Returns an error for an empty key or timestamp overflow.
    pub fn put(
        &mut self,
        key: impl Into<String>,
        kind: WebauthnCeremonyKind,
        state: State,
        subject_id: Option<String>,
        now_ms: u64,
    ) -> Result<(), WebauthnCeremonyStoreError> {
        let key = key.into();
        if key.is_empty() {
            return Err(WebauthnCeremonyStoreError::EmptyKey);
        }
        self.prune(now_ms);
        let expires_at_ms = now_ms
            .checked_add(self.ttl_ms)
            .ok_or(WebauthnCeremonyStoreError::ExpiryOverflow)?;
        self.entries.insert(
            key,
            WebauthnCeremony {
                kind,
                state,
                expires_at_ms,
                subject_id,
            },
        );
        Ok(())
    }

    /// Atomically removes and returns a matching, unexpired ceremony.
    ///
    /// The entry is removed before validation so missing, expired, mismatched, and subsequently
    /// failed cryptographic verification attempts are all single-use.
    ///
    /// # Errors
    ///
    /// Returns a stable missing, expired, or kind-mismatch category.
    pub fn consume(
        &mut self,
        key: &str,
        expected_kind: WebauthnCeremonyKind,
        now_ms: u64,
    ) -> Result<WebauthnCeremony<State>, WebauthnCeremonyStoreError> {
        let ceremony = self
            .entries
            .remove(key)
            .ok_or(WebauthnCeremonyStoreError::Missing)?;
        if ceremony.expires_at_ms <= now_ms {
            return Err(WebauthnCeremonyStoreError::Expired);
        }
        if ceremony.kind != expected_kind {
            return Err(WebauthnCeremonyStoreError::KindMismatch);
        }
        Ok(ceremony)
    }

    pub fn prune(&mut self, now_ms: u64) -> usize {
        let before = self.entries.len();
        self.entries
            .retain(|_, ceremony| ceremony.expires_at_ms > now_ms);
        before - self.entries.len()
    }

    pub fn clear(&mut self) {
        self.entries.clear();
    }

    #[must_use]
    pub fn len(&self) -> usize {
        self.entries.len()
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }
}
